eth/tracers/api_blockResult.go

```diff
@@ -0,0 +1,362 @@
+		for i, bt := range proof {
+			wrappedProof[i] = bt
+		}
+		env.Proofs[key] = wrappedProof
+	}
+	return env, nil
+}
+
+func (api *API) getBlockResult(block *types.Block, env *traceEnv) (*types.BlockResult, error) {
+	// Execute all the transaction contained within the block concurrently
+	var (
+		txs   = block.Transactions()
+		pend  = new(sync.WaitGroup)
+		jobs  = make(chan *txTraceTask, len(txs))
+		errCh = make(chan error, 1)
+	)
+	threads := runtime.NumCPU()
+	if threads > len(txs) {
+		threads = len(txs)
+	}
+	for th := 0; th < threads; th++ {
+		pend.Add(1)
+		go func() {
+			defer pend.Done()
+			// Fetch and execute the next transaction trace tasks
+			for task := range jobs {
+				if err := api.getTxResult(env, task.statedb, task.index, block); err != nil {
+					select {
+					case errCh <- err:
+					default:
+					}
+					log.Error("failed to trace tx", "txHash", txs[task.index].Hash().String())
+				}
+			}
+		}()
+	}
+
+	// Feed the transactions into the tracers and return
+	var failed error
+	for i, tx := range txs {
+		// Send the trace task over for execution
+		jobs <- &txTraceTask{statedb: env.state.Copy(), index: i}
+
+		// Generate the next state snapshot fast without tracing
+		msg, _ := tx.AsMessage(env.signer, block.BaseFee())
+		env.state.Prepare(tx.Hash(), i)
+		vmenv := vm.NewEVM(env.blockCtx, core.NewEVMTxContext(msg), env.state, api.backend.ChainConfig(), vm.Config{})
+		if _, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas())); err != nil {
+			failed = err
+			break
+		}
+		// Finalize the state so any modifications are written to the trie
+		// Only delete empty objects if EIP158/161 (a.k.a Spurious Dragon) is in effect
+		env.state.Finalise(vmenv.ChainConfig().IsEIP158(block.Number()))
+	}
+	close(jobs)
+	pend.Wait()
+
+	// If execution failed in between, abort
+	select {
+	case err := <-errCh:
+		return nil, err
+	default:
+		if failed != nil {
+			return nil, failed
+		}
+	}
+
+	return api.fillBlockResult(env, block)
+}
+
+func (api *API) getTxResult(env *traceEnv, state *state.StateDB, index int, block *types.Block) error {
+	tx := block.Transactions()[index]
+	msg, _ := tx.AsMessage(env.signer, block.BaseFee())
+	from, _ := types.Sender(env.signer, tx)
+	to := tx.To()
+
+	txctx := &Context{
+		BlockHash: block.TxHash(),
+		TxIndex:   index,
+		TxHash:    tx.Hash(),
+	}
+	// Call Prepare to clear out the statedb access list
+	state.Prepare(txctx.TxHash, txctx.TxIndex)
+
+	sender := &types.AccountWrapper{
+		Address:  from,
+		Nonce:    state.GetNonce(from),
+		Balance:  (*hexutil.Big)(state.GetBalance(from)),
+		CodeHash: state.GetCodeHash(from),
+	}
+	var receiver *types.AccountWrapper
+	if to != nil {
+		receiver = &types.AccountWrapper{
+			Address:  *to,
+			Nonce:    state.GetNonce(*to),
+			Balance:  (*hexutil.Big)(state.GetBalance(*to)),
+			CodeHash: state.GetCodeHash(*to),
+		}
+	}
+
+	tracer := vm.NewStructLogger(env.config.LogConfig)
```

Comment:
i notice that in https://github.com/scroll-tech/go-ethereum/blob/staging/eth/tracers/api.go#L896-L902, `statedb.Prepare(txctx.TxHash, txctx.TxIndex)` is put before `result, err := core.ApplyMessage(vmenv, message, new(core.GasPool).AddGas(message.Gas()))` & after `vmenv := vm.NewEVM(vmctx, txContext, statedb, api.backend.ChainConfig(), vm.Config{Debug: true, Tracer: tracer, NoBaseFee: true})`

and here you put `state.Prepare(txctx.TxHash, txctx.TxIndex)` up, will it cause any problem?

Reply:
1. Replay Tx need pre tx state(but not pre block state), the serial execution can make sure `getTxResult` get the right pre tx state.
2. `env.state.Copy()` can't copy the `env.state.Prepare(tx.Hash(), i)`, so we have to redo it in `getTxResult`.

Reply:
still prefer to move `statedb.Prepare(txctx.TxHash, txctx.TxIndex)`    between   
 result, err := core.ApplyMessage(vmenv, message, new(core.GasPool).AddGas(message.Gas())) & vmenv := vm.NewEVM(vmctx, txContext, statedb, api.backend.ChainConfig(), vm.Config{Debug: true, Tracer: tracer, NoBaseFee: true})`

Reply:
changed.